Feature decouple somd 3

Roy-Haolin-Du opened the pull request:
## Description

> Fixbug for reading the correct slurm_out_file

> EngineType enmu changed and remove the mapping dictionaries.

> Correct tmp_files name and make sure lam values enable setting from engine_config

> Integration test run locally, won't run on CI

Close https://github.com/michellab/a3fe/issues/11

## Status
All tests and format check passed
Run non-adaptive/adaptive a3fe cases successfully
- [X] Ready to go

Hi Finlay, hope you are well!
Could you please review @fjclark ?

fjclark commented on a3fe/tests/README.md:
```diff
@@ -0,0 +1,41 @@
+# A3FE Integration Tests
```
Brilliant, this looks ideal! Thanks so much for the clear documentation!

fjclark commented on a3fe/tests/__init__.py:
```diff
@@ -13,10 +13,13 @@
     RUN_SLURM_TESTS = False
+elif os.environ.get("RUN_SLURM_TESTS") is None:
+    # Default to not running SLURM tests unless explicitly enabled
+    RUN_SLURM_TESTS = False
```
I think all of the above logic can be simplified to something like
```
RUN_SLURM_TESTS = os.environ.get("RUN_SLURM_TESTS") not in [None, "False"]
```

fjclark commented on a3fe/tests/conftest.py:
```diff
@@ -125,3 +123,60 @@ def a3fe_restraint():
+
+
+@pytest.fixture(scope="session", params=[a3.EngineType.SOMD])
+def engine_type(request):
```
Very neat!

fjclark commented on a3fe/tests/test_engine_configuration.py:
```diff
@@ -1,25 +1,23 @@
 
-def test_write_config():
-    """Test that the SOMD configuration file is generated correctly."""
+def test_write_config(engine_config):
```
The writing config tests are probably specific to the engine, but you've (very nicely) set up `engine_config` so that it can be SOMD or GROMACS. Maybe directly supply a SOMD config here, and make the test `test_write_config_somd`? Feel free to keep it generic if you can though!

fjclark commented on a3fe/tests/test_engine_configuration.py:
```diff
@@ -55,15 +53,17 @@ def test_write_config():
     ],
 )
-def test_integrator_thermostat_validation(integrator, thermostat, should_pass):
+def test_integrator_thermostat_validation(
```
Again, might just be specific to SOMD?

fjclark commented on a3fe/tests/test_engine_configuration.py:
```diff
@@ -77,44 +77,44 @@ def test_integrator_thermostat_validation(integrator, thermostat, should_pass):
 
 
-def test_get_somd_config_with_extra_options():
+def test_get_somd_config_with_extra_options(engine_config):
```
Probably just pass a SOMD config here too.

fjclark commented on a3fe/tests/test_engine_configuration.py:
```diff
@@ -136,7 +136,7 @@ def test_get_somd_config_with_extra_options():
 
 
-def test_compare_with_reference_config():
+def test_compare_with_reference_config(engine_config):
```
Also here

fjclark commented on a3fe/tests/test_engine_configuration.py:
```diff
@@ -198,12 +198,12 @@ def test_compare_with_reference_config():
-def test_copy_from_existing_config():
+def test_copy_from_existing_config(engine_config):
     """Test that we can copy from an existing somd.cfg file."""
     reference_config = "/home/roy/software/deve/a3fe/a3fe/data/example_run_dir/bound/discharge/output/lambda_0.000/run_01/somd.cfg"
```
Please avoid hard-coded path so that it works on any workstation. Thanks! You can always try running tests on plato to make sure things work with different slurm settings and paths.

fjclark commented on a3fe/tests/test_run.py:
```diff
@@ -332,7 +335,7 @@ def _mock_run_process_inner(
     @staticmethod
     @pytest.fixture(scope="class")
-    def setup_calc(mock_run_process):
+    def setup_calc(mock_run_process, system_prep_config):
```
To make this generic, I think you'll need to pass in `engine_type`, which should then be supplied to `a3.Calculation` - otherwise e.g. a GROMACS system preparation config will break things.

fjclark commented on a3fe/tests/test_run_integration.py:
```diff
@@ -0,0 +1,214 @@
+Note: This file contains dedicated integration tests that are more granular than
+the integration test in test_run.py (which is currently skipped).
+
+See README.md in this directory for more information on running these tests.
```
Could you also please copy over the helpful stuff in this README.md to the docs? Thanks!

fjclark commented on a3fe/tests/test_run_integration.py:
```diff
@@ -0,0 +1,214 @@
+                except (IndexError, ValueError) as e:
+                    pytest.fail(f"Failed to parse free energy value: {str(e)}")
+
+        except FileNotFoundError as e:
```
Should the test not fail if there's a FileNotFoundError?

fjclark commented on a3fe/tests/test_run_integration.py:
```diff
@@ -0,0 +1,214 @@
+
+        calc.get_optimal_lam_vals(delta_er=2)
+
+        # Check that the old runs have been moved, and that the new lambda values are
```
Please could you check that the new lambda values are different to the old ones? Thanks!

fjclark commented on a3fe/tests/test_run_integration.py:
```diff
@@ -0,0 +1,214 @@
+        # Check that the calculation is not running
+        assert not calc.running
+
+        # Set the entire simulation time to equilibrated for analysis
```
Please don't set the equilibration time - this should be done automatically when running adaptively. Please check that the equilibration time has successfully been set. Thanks

fjclark commented on a3fe/tests/test_run_integration.py:
```diff
@@ -0,0 +1,214 @@
+        assert not calc.running
+
+        # Set the entire simulation time to equilibrated for analysis
+        for leg in calc.legs:
```
I used to have to do this before `calc.set_equil` or whatever it is was introduced. Please use this function instead.

fjclark commented on a3fe/tests/test_run_integration.py:
```diff
@@ -0,0 +1,214 @@
+                except (IndexError, ValueError) as e:
+                    pytest.fail(f"Failed to parse free energy value: {str(e)}")
+
+        except FileNotFoundError as e:
```
Again, shouldn't it fail is there's a FileNotFoundError?

fjclark reviewed (changes requested):
Thanks so much Roy, this looks great! I especially like the thorough tests and what you've done with the integration tests is really nice. Having a general engine_type fixture is a neat solution.

I've left a few small comments to address. A few things left to do before we merge into main are:

- Add extensive docs
- Possibly add a few more tests on e.g. updating config options at the calculation level and checking that the changes propagate down to the simulation level
- Possibly adding a makefile (see https://github.com/fjclark/red/blob/main/Makefile) to simplify installation and add pre-commit hooks (something like https://github.com/fjclark/red/blob/main/.pre-commit-config.yaml) to prevent us forgetting to format before we commit - I always do this. However, this isn't essential, so we can always do this later if you'd prefer.

fjclark commented on a3fe/tests/__init__.py:
```diff
@@ -14,12 +14,7 @@
-elif os.environ.get("RUN_SLURM_TESTS") is None:
-    # Default to not running SLURM tests unless explicitly enabled
-    RUN_SLURM_TESTS = False
+RUN_SLURM_TESTS = os.environ.get("RUN_SLURM_TESTS") not in [None, "False"]
```
Ah, should probably add 0 to this list (as in `RUN_SLURM_TESTS=0` should avoid running the tests, and `RUN_SLURM_TESTS=1` should run them)! Sorry I forgot

Roy-Haolin-Du commented on docs/guides.rst:
```diff
@@ -125,37 +122,44 @@ is not killed when you log out).
-you if you are trying to set a parameter that is not present in the template config file. To see a list of available options, run ``somd-freenrg --help-config``.
+    cfg = a3.SomdSystemPreparationConfig()
+    cfg.forcefields["ligand"] = "gaff2"
+    calc.setup(bound_leg_sysprep_config=cfg, free_leg_sysprep_config=cfg)
```
Hi, many thanks for your valuable request for this part of code, but I had some concerns.

  The current design good at: 
- Support flexibility to configure different legs independently like setting only one leg with specific lambda values.
`cfg.lambda_values = {a3.LegType.BOUND: {a3.StageType.RESTRAIN: [0.000, 0.125, 0.250, 0.375, 0.500, 1.000]}}`  
`a3.Leg.required_stages = {a3.LegType.BOUND: [a3.StageType.RESTRAIN], a3.LegType.FREE:[]}` 
`a3.Calculation.required_legs = [a3.LegType.BOUND] `
`calc = a3.Calculation()`
`calc.setup(bound_leg_sysprep_config=cfg,free_leg_sysprep_config=None)`
- If a single stage of one leg fails due to machine or network instability, we can reconfigure the simulation to run only that stage, avoiding redundant computation. The results can still be included in the overall analysis later on.

 The inconvenient part is:
- As you said each cfg contains information on how to set up both the bound and free legs, which doesn't make sense if it's passed only as a bound or free leg config.

Proposed Single-Parameter Design:

- Good at: More logical API with a single configuration object and avoiding the issue of settings for both legs but only applying it to one.
- Cons: Less direct flexibility for specific use cases.

From my perspective, in the current development phase, as it helps handle Slurm and interruptions caused by machine differences, maintaining the current setup is a good idea. But for future batch production, it could be simplified to use a single parameter.

fjclark replied:
Thanks for the comments Roy, I agree with your concerns. I'm still keen to change it (for the reasons below), but if you disagree, feel free to ignore this.

One thing I think we definitely should change is to get the required legs from the sys prep config file as is done for the legs (https://github.com/Roy-Haolin-Du/a3fe/blob/81b502176c963dcf02301dad35618c6526fb5358/a3fe/run/leg.py#L261). This means that we don't need the `required_legs` class attribute, ensures that the code works consistently (always reads from the preparation config) and is in keeping with the single source of information principle (the setup config knows everything about how things should be set up, and we always consult it). Please could you:

- Delete the `required_legs` class attribute
- Add a `required_legs` property to the steup config, which will look very similar to https://github.com/Roy-Haolin-Du/a3fe/blob/81b502176c963dcf02301dad35618c6526fb5358/a3fe/configuration/system_prep_config.py#L128
- Read the required legs from the sys prep config instead of the class attribute. 

Then, if you only wanted to run bound restrain, you could just do:

```
cfg.lambda_values = {a3.LegType.BOUND: {a3.StageType.RESTRAIN: [0.000, 0.125, 0.250, 0.375, 0.500, 1.000]}}
calc = a3.Calculation()
calc.setup(sysprep_config=cfg) # Assuming that we switch to one config
```
with no need to explicity do `a3.Calculation.required_legs = [a3.LegType.BOUND] `. I think the line `a3.Leg.required_stages = {a3.LegType.BOUND: [a3.StageType.RESTRAIN], a3.LegType.FREE:[]}` in your code above is already not required based on your changes - you already read this from the sysprep config (e.g. https://github.com/Roy-Haolin-Du/a3fe/blob/81b502176c963dcf02301dad35618c6526fb5358/a3fe/run/leg.py#L407). 

So supplying only a single config would give you complete control over the lambda windows and what stages/ legs get run, although you're right that this would stop you configuring e.g. different time steps between stages. What specific options (other than lambda windows/ stages/ legs run) do you think would be useful to set on a per-leg level? If there are only one or two, we could always change their values in the config to dicts, e.g. timestep = {a3.LegType.BOUND : 2, a3.LegType.FREE : 4} rather than duplicating the entire config.

Thanks!

Roy-Haolin-Du commented:
Description

- Simplify env setup in Makefile, pre-commit-hooks and CI configuration
- Refactor calc.setup : `sysprep_config` setting both bound and free leg
- Added extensive docs and modified all test files to make sure meeting requirements and working

All tests and format check passed
Run non-adaptive/adaptive a3fe cases successfully

Hi Finlay, hope you are well. 
I know you’ve got a busy week with the group meeting and your viva,
Just take a look whenever you have a moment please, no rush!
@fjclark 

Thanks so much.

fjclark commented on Makefile:
```diff
@@ -1,23 +1,18 @@
-	# Install development tools
-	$(CONDA_ENV_RUN) pip install -e .
+	mamba create -y --name $(PACKAGE_NAME)
+	mamba env update --name $(PACKAGE_NAME) --file devtools/conda-envs/test_env.yaml
```
Thanks, this is nice and simple

fjclark commented on Makefile:
```diff
@@ -52,4 +47,11 @@ clean:
+ifndef VERSION
+	$(error VERSION is not set)
+endif
+	$(CONDA_ENV_RUN) mike deploy --push --update-aliases $(VERSION)
```
Nice, can the docs now be deployed with mike? That's great if we now have versioned docs - before I did it badly so only the last docs were accessible.

fjclark commented on a3fe/run/calculation.py:
```diff
@@ -143,7 +141,7 @@ def _validate_input(self) -> None:
         for prep_stage in reversed(_PreparationStage):
             files_absent = False
-            for leg_type in Calculation.required_legs:
+            for leg_type in [_LegType.FREE, _LegType.BOUND]:
```
Could we avoid hard-coding this? I'm imagining the case where someone only wants to run e.g. the free leg for some reason. I know that it's hard because we don't yet have the sys prep config, which would contain this information. Some options: 1) supply the sys prep config to __init__ instead of setup 2) Only validate when someone attempts to run setup. What do you think?

fjclark commented on a3fe/tests/README.md:
````diff
@@ -6,23 +6,26 @@ This directory contains integration tests for A3FE, which are designed to run in
-```
+   ```bash
+   cd your/path/a3fe
+   RUN_SLURM_TESTS=1 pytest a3fe/tests --run-integration -v
````
Maybe you could also mention they can now run "make test-integration" - thanks for adding that!

fjclark commented on a3fe/tests/README.md:
````diff
@@ -6,23 +6,26 @@ This directory contains integration tests for A3FE, which are designed to run in
+   Or to explicitly skip integration tests (same as deafult):
    ```bash
-   RUN_SLURM_TESTS=1 pytest a3fe/tests/test_run_integration.py::TestSlurmIntegration::test_slurm_calculation_setup --run-integration -v
+   RUN_SLUEM_TESTS=0 pytest a3fe/tests --run-integration -v
````
Small typo "SLUEM". Also, should --run-integration be removed to avoid confusion? My understanding is that this says "run the integration tests" but RUN_SLURM_TESTS=0 stops them running. In the future, these commands might do different things.

fjclark commented on a3fe/tests/test_run_integration.py:
```diff
@@ -202,11 +202,39 @@ def test_slurm_adaptive_run(self, slurm_calc_adaptive, system_prep_config):
+            for stage_name in LEGS_WITH_STAGES[leg_name]:
+                # Check that the check_equil_multiwindow_paired_t.txt file exists
+                leg_path = os.path.join(calc.base_dir, leg_name)
+                stage_path = os.path.join(leg_path, stage_name)
```
This looks great. Just commenting on something I've done badly - I've mixed the using strings as paths with os.path.join and using pathlib.Path throughout the code. I think pathlib.Path is nicer and if I wrote from scratch I would only use pathlib.Path. Just mentioning for awareness though - this isn't something we really need to fix.

fjclark commented on docs/CHANGELOG.rst:
```diff
@@ -5,10 +5,18 @@ Change Log
 
+**Migration Guide**:
+In previous versions, users needed to configure run_somd.sh and template_config.cfg files to run calculations. 
+However, with version 0.4.0, these files are no longer needed. Instead, configuration objects (SlurmConfig, SomdConfig, and SomdSystemPreparationConfig) are passed directly to the Calculation object.
```
Brilliant, thanks! Maybe add a link to where the docs show an example of doing this?

fjclark commented on docs/guides.rst:
```diff
@@ -130,14 +130,19 @@ For example, to change the timestep, create a ``SomdConfig`` object and pass it
     calc = a3.Calculation()
-    calc.engine_config.timestep = 2.0
+    # Works if the calculation has not been setup yet
+    calc.engine_config.timestep = 2.0 # fs
```
Thanks very much for pointing this out in the docs - this is a super important point. I feel like this may still cause some confusion if people don't read the docs properly. I wonder if we could warn the user if they edit the config directly when the calc has already been setup? Do you have any ideas? Don't worry if not as it's already way better than it used to be! This is fine, just trying to minimise confusion. Might be worth reminding users why this is required - just saying because the configs for the legs etc also need updated

fjclark commented on docs/guides.rst:
```diff
@@ -130,14 +130,19 @@ For example, to change the timestep, create a ``SomdConfig`` object and pass it
+
 
 To see a complete list of available configuration options, run ``somd-freenrg --help-config``.
+or inspect the ``a3.SomdConfig`` class in the code.
```
Could maybe link to the API reference for this class

fjclark commented:
Hi @Roy-Haolin-Du,

Thanks very much for the changes, and sorry it took me so long to review - have been super busy recently. This looks great and I've just left a couple of very small and pedantic suggestions. It's pretty much ready to go! I'd suggest this route to release, if good with you:

 - You address the minor changes above, when you have time and if you agree with the comments
 - I'll double-check your changes, then have a quick play around with the new code on ishikawa, when you have some time between calculations - if that's ok (I don't have access to any SLURM workstations at the moment)?
 - Merge and release!
 - You create a zenodo doi for the release including you as an author and stick it on the readme so that people can cite you if they use the code

Does that sound good?

Sorry again for the mega delay, and apologies in advance for more slow replies - I've got quite a few holidays and other work on til mid June, but should be less busy then!

fjclark commented:
I've just been testing the new code - this is mostly great!

One minor comment - in the Makefile, you do `mamba env update --name $(PACKAGE_NAME) --file devtools/conda-envs/test_env.yaml`, (which installs the dev/ test dependencies, including GROMACS, which messes up using the local compiled version with CUDA). Could you please install from the top-level environment.yaml instead please? You could always move this into `devtools/conda-envs` and call it e.g. `base.yaml` or similar. Thanks!

fjclark commented:
Another thing which we should fix (I think this came up with Silvia - maybe we've fixed this already...) - I should have been more careful about modifying paths:

If there is `input` in the path name, a3fe will change this to `output` e.g. I was running in `~/research/polaris_retrospective/asma_input_testing_160625/test_a3fe_calc/bound/vanish` but this got changed to `~/research/polaris_retrospective/asma_output_testing_160625/test_a3fe_calc/bound/vanish` for the output.

fjclark commented:
Ah, forgot you already fixed this in https://github.com/michellab/a3fe/pull/37/files! Thanks! We just need to merge main into these changes. :)

Roy-Haolin-Du commented:
> I've just been testing the new code - this is mostly great!
> 
> One minor comment - in the Makefile, you do `mamba env update --name $(PACKAGE_NAME) --file devtools/conda-envs/test_env.yaml`, (which installs the dev/ test dependencies, including GROMACS, which messes up using the local compiled version with CUDA). Could you please install from the top-level environment.yaml instead please? You could always move this into `devtools/conda-envs` and call it e.g. `base.yaml` or similar. Thanks!

Hi, Finlay, thanks so much for your valuable suggestion.

For my perspective, due to CI needs GROMACS package for dependency testing, however, local developers already had CUDA GROMACS and wanting to test tools but avoid conda GROMACS conflicting. 
For regular users, which only need general run dependencies, no development tools. 

Therefore, to meet the followings:
1. Regular users : `mamba env create -f environment.yaml` as shown in README.
2. Developers with local GROMACS: `make env `    ## create development environment without GROMACS from `base_env.yaml`
3. CI + some developers without GROMACS: `make env-full`  ##create full development environment including GROMACS from `test_env.yaml`

Each user type gets exactly what they need. 

Do you think this is a convenient and reasonable approach? Many thanks!

fjclark commented:
Thanks so much for this Roy! And sorry I took so long to get round to having a look. I've had a quick play around and checked your changes. The code looks great and I don't think you need any more input from me! This is a big improvement to `a3fe` and I'm super excited about it. Thanks especially for investing time in updating the docs and adding more tests :)

To release, could you make these small tweaks/ tests please? No need to get more input from me -- release once you've done the below and are happy! 

## Some tiny code changes
Sorry, there's always some...

- Make sure `set_simfile_option` is removed everywhere, because we no longer have simfiles. Also please double-check that all redundant functionality is removed.
- Make sure that the pydantic validation is working. At the moment, I can do `a3.SomdConfig(cutoff_distance=-4)` and no error is raised, and I can also do `calc.update_engine_config_option("cutoff_distance",-4)` and no error is raised. To fix this, you need to add 
```
    model_config = {
        "populate_by_name": True, 
        "validate_assignment": True,
    }
```
inside `_EngineConfig`. The first line ensures the first example fails and the second line ensures the second example fails. Please check that `"validate_assignment": True,` is set in all pydantic configs so that they can't be messed up after creation. I'd advise adding a test that fails with the code in it's current state, too (e.g. even just  `a3.SomdConfig(cutoff_distance=-4)`).

## Some other minor things to do before releasing

- Change the copyright notice to 2023-2025 and include your name
- Check that the test coverage hasn't dropped too much (ideally increased) from the current 73
- Merge into `michellab:feature-decouple-somd-3` then merge in the changes from the main branch. Test everything again at this stage to make sure nothing has been broken. The merge might be painful...

## Release!

- Merge into main and release as normal! It's a big change so a nice summary of the updates/ pointing to bits of the docs on the release would be super helpful.
- Create a Zenodo DOI with both of us as authors. See https://zenodo.org/records/7433270 for inspiration
- Update the readme to ask people to cite the Zenodo as well as the paper if they use it. Also, you'll probably want to re-add the git version label as 0.4.0 after this commit so that when people install they don't end up with the .dirty..... in the version number.

Again, this looks great to me so no need to get me to re-review unless you're uncertain about something (in which case I'm very happy to take a look)!

Thanks again Roy!

fjclark commented:
Ah also, please could you:

- When you merge, close the issues which this PR solves e.g. https://github.com/michellab/a3fe/issues/20
- When telling people what to cite, mention that this is all basically sire and BioSimSpace and as well as the new doi and original a3fe paper, they should check out/cite Sire (https://pubs.aip.org/aip/jcp/article/160/20/202503/3295373) and BioSimSpace (https://livecomsjournal.org/index.php/livecoms/article/view/v5i1e2375 and https://joss.theoj.org/papers/10.21105/joss.01831). 

Thanks!

Roy-Haolin-Du commented:
Hi @fjclark 😄

Hope you are well and sorry to trouble you~
I’m a bit uncertain about the adaptive A3FE. 

In the original defaults we used the number of lamba windows: bound–vanish = 36 lambdas, restrain = 5, discharge = 8. In the current decoupleSOMD A3FE, the bound–vanish stage only has 20 windows (fewer overall).

In my tests, for the same ligand, the adaptive run need to take longer for decoupleSOMD. If we use a denser initial lambda spacing allows us to determine the lambda positions more accurately. But for lower default lambda density, the error propagation seems to amplify from SEM to normalised_sem_dg, which then triggers more ns to be run.
By contrast, the get_opt_lam_val step uses 0.1 ns and repeats only once, so it’s very fast.

Could we spend a bit more time in get_opt_lam_val to obtain a more suitable lambda schedule. This should reduce error amplification later and save substantial time during the adaptive stage, which we should use more denser lambda windows as default like before.
What do you think about this?

Thanks!
Roy

fjclark commented:
Hi Roy, doing well thanks -- hope you are too!

Firstly, you are now the number one A3FE user, so please feel free to use the defaults that you think are best!

> But for lower default lambda density, the error propagation seems to amplify from SEM to normalised_sem_dg, which then triggers more ns to be run

Just to make sure I understand -- do you mean that when we use the new (wider) defaults for `get_opt_lam_val`, this results in more widely spaced optimised lambda windows, which then results in the allocation of more sampling time? (I would expect more sampling time to be allocated for more widely spaced windows for the bound leg, as shown in table 2 in https://pubs.acs.org/doi/full/10.1021/acs.jctc.4c00806, as the central limit theorem isn't satisfied). Could you post the overlap matrices/ window spacings you get in each case, please?

But, again, please specify the defaults which you find work best!

Thanks,
Finlay
